File: task1.go
package main

import (
	"fmt"
	"math"
)

// Точность
const epsilon = 0.0001

// Исходная функция
func f(x float64) float64 {
	return math.Sin(x+1) - x - 1
}

// Производная
func df(x float64) float64 {
	return math.Cos(x+1) - 1
}

// Функция phi(x)
func phi(x float64) float64 {
	return math.Sin(x+1) - 1
}

// Метод половинного деления
func halfDivision(a float64, b float64) {

	fmt.Println("\nМетод половинного деления")
	fmt.Println("--------------------------------------------------")

	fmt.Printf("%5s%15s%15s%15s\n",
		"N", "a_n", "b_n", "b_n-a_n")

	n := 0
	var c float64

	for (b - a) >= epsilon {

		c = (a + b) / 2.0

		fmt.Printf("%5d%15.6f%15.6f%15.6f\n",
			n, a, b, (b - a))

		if math.Abs(f(c)) < epsilon {
			break
		}

		if f(a)*f(c) < 0 {
			b = c
		} else {
			a = c
		}

		n++
	}

	fmt.Printf("\nКорень: %.10f\n", c)
	fmt.Printf("f(x) = %.10f\n", f(c))
	fmt.Printf("Количество итераций: %d\n", n)
}

// Метод Ньютона
func newtonMethod(x0 float64) {

	fmt.Println("\nМетод Ньютона")
	fmt.Println("--------------------------------------------------")

	fmt.Printf("%5s%15s%15s%15s\n",
		"N", "x_n", "x_n+1", "|dx|")

	x := x0
	var xNext float64

	n := 0

	for {

		xNext = x - f(x)/df(x)

		fmt.Printf("%5d%15.8f%15.8f%15.8f\n",
			n, x, xNext, math.Abs(xNext-x))

		if math.Abs(xNext-x) < epsilon {
			break
		}

		x = xNext
		n++
	}

	fmt.Printf("\nКорень: %.10f\n", xNext)
	fmt.Printf("f(x) = %.10f\n", f(xNext))
	fmt.Printf("Количество итераций: %d\n", n)
}

// Метод простых итераций
func simpleIteration(x0 float64) {

	fmt.Println("\nМетод простых итераций")
	fmt.Println("--------------------------------------------------")

	fmt.Printf("%5s%15s%15s%15s\n",
		"N", "x_n", "x_n+1", "|dx|")

	x := x0
	var xNext float64

	n := 0

	for {

		xNext = phi(x)

		fmt.Printf("%5d%15.8f%15.8f%15.8f\n",
			n, x, xNext, math.Abs(xNext-x))

		if math.Abs(xNext-x) < epsilon {
			break
		}

		x = xNext
		n++

		if n > 100 {
			break
		}
	}

	fmt.Printf("\nКорень: %.10f\n", xNext)
	fmt.Printf("f(x) = %.10f\n", f(xNext))
	fmt.Printf("Количество итераций: %d\n", n)
}

// Главная функция
func main() {

	fmt.Println("Решение уравнения:")
	fmt.Println("sin(x + 1) - x - 1 = 0")

	fmt.Println("\nОтделение корня:")
	fmt.Println("x принадлежит [-1; 0]")

	// Метод половинного деления
	halfDivision(-1.0, 0.0)

	// Метод Ньютона
	newtonMethod(-0.5)

	// Метод простых итераций
	simpleIteration(-0.5)

	fmt.Println("\nСравнение методов:")
	fmt.Println("Метод Ньютона сходится быстрее остальных.")
}

File: task1.cpp
#include <iostream>
#include <cmath>
#include <iomanip>

using namespace std;

const double epsilon = 0.0001;

// Исходная функция
double f(double x) {
    return sin(x + 1) - x - 1;
}

// Производная
double df(double x) {
    return cos(x + 1) - 1;
}

// Функция phi(x)
double phi(double x) {
    return sin(x + 1) - 1;
}

// Метод половинного деления
void halfDivision(double a, double b) {

    cout << "\nМетод половинного деления\n";
    cout << "--------------------------------------------------\n";

    cout << setw(5)  << "N"
         << setw(15) << "a_n"
         << setw(15) << "b_n"
         << setw(15) << "b_n-a_n"
         << endl;

    int n = 0;
    double c;

    while ((b - a) >= epsilon) {

        c = (a + b) / 2.0;

        cout << setw(5)  << n
             << setw(15) << fixed << setprecision(6) << a
             << setw(15) << b
             << setw(15) << (b - a)
             << endl;

        if (fabs(f(c)) < epsilon)
            break;

        if (f(a) * f(c) < 0)
            b = c;
        else
            a = c;

        n++;
    }

    cout << "\nКорень: " << setprecision(10) << c << endl;
    cout << "f(x) = " << f(c) << endl;
    cout << "Количество итераций: " << n << endl;
}

// Метод Ньютона
void newtonMethod(double x0) {

    cout << "\nМетод Ньютона\n";
    cout << "--------------------------------------------------\n";

    cout << setw(5)  << "N"
         << setw(15) << "x_n"
         << setw(15) << "x_n+1"
         << setw(15) << "|dx|"
         << endl;

    double x = x0;
    double x_next;

    int n = 0;

    while (true) {

        x_next = x - f(x) / df(x);

        cout << setw(5)  << n
             << setw(15) << fixed << setprecision(8) << x
             << setw(15) << x_next
             << setw(15) << fabs(x_next - x)
             << endl;

        if (fabs(x_next - x) < epsilon)
            break;

        x = x_next;
        n++;
    }

    cout << "\nКорень: " << setprecision(10) << x_next << endl;
    cout << "f(x) = " << f(x_next) << endl;
    cout << "Количество итераций: " << n << endl;
}

// Метод простых итераций
void simpleIteration(double x0) {

    cout << "\nМетод простых итераций\n";
    cout << "--------------------------------------------------\n";

    cout << setw(5)  << "N"
         << setw(15) << "x_n"
         << setw(15) << "x_n+1"
         << setw(15) << "|dx|"
         << endl;

    double x = x0;
    double x_next;

    int n = 0;

    do {

        x_next = phi(x);

        cout << setw(5)  << n
             << setw(15) << fixed << setprecision(8) << x
             << setw(15) << x_next
             << setw(15) << fabs(x_next - x)
             << endl;

        if (fabs(x_next - x) < epsilon)
            break;

        x = x_next;
        n++;

    } while (n < 100);

    cout << "\nКорень: " << setprecision(10) << x_next << endl;
    cout << "f(x) = " << f(x_next) << endl;
    cout << "Количество итераций: " << n << endl;
}

int main() {

    setlocale(LC_ALL, "Russian");

    cout << "Решение уравнения:\n";
    cout << "sin(x + 1) - x - 1 = 0\n";

    cout << "\nОтделение корня:\n";
    cout << "x принадлежит [-1; 0]\n";

    halfDivision(-1.0, 0.0);

    newtonMethod(-0.5);

    simpleIteration(-0.5);

    cout << "\nСравнение методов:\n";
    cout << "Метод Ньютона сходится быстрее остальных.\n";

    return 0;
}
